### ISSUE TITLE
Make HasKernel public or introduce geo::Float trait?

### ISSUE THREAD

**Issue by michaelkirk** [georust/geo#577]
A lot of our algorithms depend on Float + HasKernel. 

When downstream users want to implement methods that are generic across these same types they are stymied because HasKernel is not public outside of the geo crate.

We could make HasKernel public, and then they could implement for Float+HasKernel, but another option (and my current preference) would be to export a combined trait like:

`pub trait geo::Float: geo::CoordinateType + geo::HasKenel + num_traits::Float { }`

Note, I came across this issue while working on https://github.com/urschrei/polylabel-rs/pull/6 which is a downstream user of geo.

**public `HasKernel` approach looks like:**

geo: https://github.com/georust/geo/compare/mkirk/pub-has-kernel?expand=1

downstream geo user (polylabel): https://github.com/urschrei/polylabel-rs/compare/master...michaelkirk:mkirk/update-geo?expand=1

**`geo::Float` approach (my preference) looks like:**

geo: https://github.com/georust/geo/compare/mkirk/geo-float?expand=1
downstream geo user (polylabel): https://github.com/urschrei/polylabel-rs/compare/master...michaelkirk:mkirk/update-geo-2?expand=1

The reason I prefer the `geo::Float` approach is:

1. it's a bit tidier from the client perspective (rather than `num_traits::Float+geo::HasKernel` they just have  `geo::Float` (in fact we could tidy up the geo code base a bit with this)

2. The `HasKernel` type is a bit unintuitive for a casual reader. I think the concept of a Float will be more approachable for people just "trying to make things work".

It's very possible there's a better approach, or that I'm missing something in my analysis, so curious what other folks think.

**Comment by urschrei** [comment]
I'm in favour of the new public geo::Float trait approach. We can clearly and succinctly document what it's for and what it does in a few lines, and spare downstream users the (entirely necessary internal) complexity of HasKernel IMO. @rmanoka what do you think?

**Comment by rmanoka** [comment]
TL;DR:  I suggest we do both: create a "alias" trait `geo::Float`, as well as make `HasKernel` public.  Actually just making the alias trait public, would still be unusable as the end-user would anyway need to implement `HasKernel` to implement the alias trait on a new type.

Comp. geom. algorithms involve two concepts:  constructions (such as finding mid point of a line segment), and predicate (such as checking if point lies on / above / below a line segment).   In general, constructions are not exact unless we use very complicated exact-arithmetic types (eg.  mid point of a line-segment with any finite precision coords. like `i64`, `f64` is inexact due to the finite precision).  Otoh, predicates can be exact even for finite precision coords. and in fact the Shewchuk robust predicates do exactly that for `f64`.

Our `HasKernel` trait is meant to abstract out the robust predicates, so we can provide implementations of the algos. for integer, float, and possibly other types.  The `Float` trait is used where we need constructions, and thus may be in-exact.  So, it seems useful to make `HasKernel` public, as well as introduce new `geo::Float` as a synonym for `HasKernel + num::Float`.  Note that `HasKernel` actually has default implementations of all it's methods, so the user can get away with not implementing anything if they do not care about robustness.

The technicality of `HasKernel` is unfortunate, but I don't know how to circumvent it if we want to support robust predicates for generic scalar types.  In fact, other comp. geom. libs like spade, and CGAL also use this concept of a "Kernel" (and was the inspiration for our impl. too).  Our lib. has a key advantage over JTS/geos in that it supports generic scalar types, while also aiming to be robust and conform to standards such as SFS.  IMO, we should keep it that way.

I do agree that we should improve the docs for `HasKernel`, and may be provide an implementation guide on new types.  Should we track that as a new issue?  I could take a shot at it early next year :)

**Comment by michaelkirk** [comment]
>  just making the alias trait public, would still be unusable as the end-user would anyway need to implement HasKernel to implement the alias trait on a new type

That's a good point! Though making a `geo::Float` alias trait public is sufficient for my purposes here: https://github.com/urschrei/polylabel-rs/pull/6, it wouldn't be sufficient for someone wanting to use types other than those we've already `has_kernel!`'d.

> The technicality of HasKernel is unfortunate

To be clear, I think the `HasKernel` stuff makes a lot of sense as is, and I'm glad you did it that way. 

Only based on the name, I'm hoping non-experts will have a better intuition that "The thing you're probably looking for is `geo::Float`" rather than "The thing you're probably looking for is `geo::HasKernel + num_traits::Float`"

At the point people want to do geometry with exotic numeric types, then it's more reasonable for us to expect them to understand exactly what the `geo::HasKernel` vocabulary is for, until then, the 99% of people who are just doing stuff like: https://github.com/urschrei/polylabel-rs/pull/6 will see that we have our own weird Float trait and that they need to use it if they want to be generic.

So, to recap, I've been persuaded by @rmanoka that we should both add a public `geo::Float` trait and also make `geo::HasKernel` public.

**Comment by michaelkirk** [comment]
> I do agree that we should improve the docs for HasKernel, and may be provide an implementation guide on new types. Should we track that as a new issue? I could take a shot at it early next year :)

To address this and reiterate: my concern was not that HasKernel was unnecessarily complicated (though better docs are always welcome!), but that I'm hoping to avoid exposing people to it directly unless they are doing stuff with exotic number types.

**Comment by urschrei** [comment]
> To address this directly and reiterate: my concern was not that HasKernel was unnecessarily complicated

I think this was my doing, and just to reiterate what @michaelkirk said: the HasKernel mechanism is an excellent design and I'm delighted with its implementation! The fact that it's complex is inherent to its intended aim, and I think I speak for everyone when I say that it's solved the problem very well.

**Comment by michaelkirk** [comment]
This was done in https://github.com/georust/geo/pull/583
